### PR TITLE
release chart

### PR DESCRIPTION
**Pull request by kalantar**

### CODE REVIEW

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+              # primary version
+              {{- $v := (index $versions 0) }}
+              - destination:
+                  host: {{ template "svc.name" $v}}.{{ $APP_NAMESPACE }}.svc.cluster.local
+                  port:
+                    number: {{ $v.port }}
+                {{- if gt (len $versions) 1 }}
+                {{ `{{- if gt (index .Weights 1) 0 }}` }}
+                weight: {{ `{{ index .Weights 0 }}` }}
+                {{ `{{- end }}`}}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
                {{ `{{- end }}` }}
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                {{- end  }}
+                headers: 
+                  response:
+                    add:
+                      app-version: {{ template "svc.name" $v}}
+              # other versions
+              {{- range $i, $v := (rest $versions) }}
+              {{ `{{- if gt (index .Weights ` }}{{ print (add1 $i) }}{{ `) 0 }}`}}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
              {{ `{{- if gt (index .Weights ` }}{{ print (add1 $i) }}{{ `) 0 }}` }}
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+              - destination:
+                  host: {{ template "svc.name" $v}}.{{ $APP_NAMESPACE }}.svc.cluster.local
+                  port:
+                    number: {{ $v.port }}
+                weight: {{ `{{ index .Weights `}}{{ print (add1 $i) }}{{` }}`}}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
                weight: {{ `{{ index .Weights ` }}{{ print (add1 $i) }}{{ ` }}` }}
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                headers:
+                  response:
+                    add:
+                      app-version: {{ template "svc.name" $v}}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
                      app-version: {{ template "svc.name" $v }}
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+              {{ `{{- end }}`}}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
              {{ `{{- end }}` }}
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            - {{ $APP_NAME }}.{{ $APP_NAMESPACE }}
+            - {{ $APP_NAME }}.{{ $APP_NAMESPACE }}.svc
+            - {{ $APP_NAME }}.{{ $APP_NAMESPACE }}.svc.cluster.local
+            http:
+            - name: {{ $APP_NAME }}
+              route:
+              # primary version
+              {{- $v := (index $versions 0) }}
+              - destination:
+                  host: {{ template "svc.name" $v}}.{{ $APP_NAMESPACE }}.svc.cluster.local
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
                  host: {{ template "svc.name" $v }}.{{ $APP_NAMESPACE }}.svc.cluster.local
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  strSpec: |
+    versions: 
+    {{- range $i, $v := $versions }}
+    - resources:
+      - gvrShort: svc
+        name: {{ template "svc.name" $v}}
+        namespace: {{ template "svc.namespace" $v}}
+      - gvrShort: deploy
+        name: {{ template "deploy.name" $v}}
+        namespace: {{ template "deploy.namespace" $v}}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.routemap.tpl]:
```suggestion
        name: {{ template "svc.name" $v }}
        namespace: {{ template "svc.namespace" $v }}
      - gvrShort: deploy
        name: {{ template "deploy.name" $v }}
        namespace: {{ template "deploy.namespace" $v }}
```

File: charts/release/templates/_deployment-istio.blue-green.tpl [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{{- define "env.deployment-istio.blue-green" }}
+
+{{- /* prepare versions for simpler processing */}}
+{{- $versions := include "normalize.versions.deployment" . | mustFromJson }}
+
+{{- /* weight-config ConfigMaps */}}
+{{- range $i, $v := $versions }}
+    {{ include "configmap.weight-config" $v }}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.blue-green.tpl:8]:
```suggestion
{{ include "configmap.weight-config" $v }}
```

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl (hunk 7 above)
Comment (Alan-Cha) [reply charts/release/templates/_deployment-istio.blue-green.routemap.tpl]: This seems to be a common problem

File: charts/release/templates/_deployment-istio.canary.routemap.tpl [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    - resources:
+      - gvrShort: svc
+        name: {{ template "svc.name" $v}}
+        namespace: {{ template "svc.namespace" $v}}
+      - gvrShort: deploy
+        name: {{ template "deploy.name" $v}}
+        namespace: {{ template "deploy.namespace" $v}}
+    {{- end }}
+
+TBD: _deployment-istio.canary.routemap.tpl
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment-istio.canary.routemap.tpl]: Should this be here? Or at least exist as a comment?

File: charts/release/templates/_deployment.tpl [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+---
+{{- end }} {{- /* range $i, $v := $versions */}}
+
+{{- /* routemap (and other strategy specific objects) */}}
+{{- if not .Values.application.strategy }}
+{{ include "env.deployment-istio.none" . }}
+{{- else if eq "none" .Values.application.strategy }}
+{{ include "env.deployment-istio.none" . }}
+{{- else }}
+{{- printf "unknown or invalid application strategy (%s) for environment (%s)" .Values.application.strategy .Values.environment | fail }}
```
Comment (Alan-Cha) [inline charts/release/templates/_deployment.tpl:22]: It's not clear to me what's going on here. Why is the only acceptable strategy here `none`?

Comment (kalantar) [reply charts/release/templates/_deployment.tpl:22]: This is the scenario where we have an application made up of a Deployment and a Service but do not have Istio. As a consequence, we cannot implement blue-green or canary. Only none is valid. This is what is used for A/B/n.

File: charts/release/templates/_deployment-istio.blue-green.routemap.tpl (hunk 7 above)
Comment (kalantar) [reply charts/release/templates/_deployment-istio.blue-green.routemap.tpl]: very common. i've fixed them all.

File: charts/release/templates/_deployment-istio.canary.routemap.tpl (hunk 9 above)
Comment (kalantar) [reply charts/release/templates/_deployment-istio.canary.routemap.tpl]: added the template for this

File: charts/release/templates/_helpers.tpl [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+  {{- /* Assumption: .Values.application is valid */}}
+  {{- $metadata := dict }}
+  {{- if .Values.application.metadata }}
+  {{- $metadata = merge $metadata .Values.application.metadata }}
+  {{- end }} {{- /* if .Values.application.metadata */}}
+
+  {{- $APP_NAME := (include "application.name" .) }}
+  {{- $APP_NAMESPACE := (include "application.namespace" .) }}
+
+  {{- $defaultMatch := ternary (list (dict "headers" (dict "traffic" (dict "exact" "test")))) (dict) (eq .Values.application.strategy "canary") }}
```
Comment (Alan-Cha) [inline charts/release/templates/_helpers.tpl:76]: A comment to describe what is happening here would be appreciated

Comment (kalantar) [reply charts/release/templates/_helpers.tpl:76]: done

File: charts/release/templates/_helpers.tpl [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+  {{- $APP_NAME := (include "application.name" .) }}
+  {{- $APP_NAMESPACE := (include "application.namespace" .) }}
+
+  {{- /* set default match for canary use cases: "traffic: test" */}}
+  {{- /* this is arbitrary but enables trying it out quickly " */}}
+  {{- $defaultMatch := ternary (list (dict "headers" (dict "traffic" (dict "exact" "test")))) (dict) (eq .Values.application.strategy "canary") }}
+
+  {{- $normalizedVersions := list }}
+  {{- range $i, $v := .Values.application.versions -}}
+    {{- $version := merge $v }}
```
Comment (Alan-Cha) [inline charts/release/templates/_helpers.tpl:80]: What is the purpose of this?

File: charts/release/templates/_helpers.tpl [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    {{- $VERSION_NAME := (printf "%s-%d" $APP_NAME $i) }}
+    {{- if (and $v.metadata $v.metadata.name) }}
+    {{- $VERSION_NAME = $v.metadata.name }}
+    {{- end }} {{- /* if (and $v.metadata $v.metadata.name) */}}
+    {{- $VERSION_NAMESPACE := $APP_NAMESPACE }}
+    {{- if (and $v.metadata $v.metadata.namespace) }}
+    {{- $VERSION_NAMESPACE = $v.metadata.namespace }}
+    {{- end }} {{- /* if (and $v.metadata $v.metadata.namespace) */}}
+    {{- $version := merge $v }}
```
Comment (Alan-Cha) [inline charts/release/templates/_helpers.tpl:89]: Similarly here...

File: charts/release/templates/_helpers.tpl (hunk 12 above)
Comment (Alan-Cha) [reply charts/release/templates/_helpers.tpl:80]: Is it necessary to use the merge function?

File: controllers/routemap.go [hunk 14]
```diff
@@ -473,12 +492,12 @@ func getObservedGeneration(obj *unstructured.Unstructured, condition map[string]
...
-		err := errors.New("routemap CM is not immutable")
-		log.Logger.Error(err)
-		return err
-	}
+	// // routemap CM must be immutable
+	// if confMap.Immutable == nil || !(*confMap.Immutable) {
+	// 	err := errors.New("routemap CM is not immutable")
+	// 	log.Logger.Error(err)
+	// 	return err
+	// }
```
Comment (Alan-Cha) [inline controllers/routemap.go]: Currently, this function isn't really being used?

Comment (kalantar) [reply controllers/routemap.go]: It is not used. Removed commented out code.

**Review by Alan-Cha — APPROVED** [review]

**Pull request merged**